src/toil/provisioners/aws/awsProvisioner.py

```diff
@@ -216,32 +246,44 @@ def createClusterSettings(self):
         # constructor.
         assert self._zone is not None
 
-    def readClusterSettings(self):
+    def readClusterSettings(self) -> None:
         """
         Reads the cluster settings from the instance metadata, which assumes
         the instance is the leader.
         """
-        instanceMetaData = get_instance_metadata()
-        ec2 = self.aws.boto2(self._region, 'ec2')
-        instance = ec2.get_all_instances(instance_ids=[instanceMetaData["instance-id"]])[0].instances[0]
+        from ec2_metadata import ec2_metadata
+        boto3_ec2 = self.aws.client(self._region, 'ec2')
+        instance: InstanceTypeDef = boto3_ec2.describe_instances(InstanceIds=[ec2_metadata.instance_id])["Reservations"][0]["Instances"][0]
         # The cluster name is the same as the name of the leader.
-        self.clusterName = str(instance.tags["Name"])
+        self.clusterName: str = "default-toil-cluster-name"
+        for tag in instance["Tags"]:
+            if tag.get("Key") == "Name":
+                self.clusterName = tag["Value"]
         # Determine what subnet we, the leader, are in
-        self._leader_subnet = instance.subnet_id
+        self._leader_subnet = instance["SubnetId"]
         # Determine where to deploy workers.
         self._worker_subnets_by_zone = self._get_good_subnets_like(self._leader_subnet)
 
-        self._leaderPrivateIP = instanceMetaData['local-ipv4']  # this is PRIVATE IP
-        self._keyName = list(instanceMetaData['public-keys'].keys())[0]
-        self._tags = {k: v for k, v in self.getLeader().tags.items() if k != _TAG_KEY_TOIL_NODE_TYPE}
+        from boto.utils import get_instance_metadata
+
+        self._keyName = instance["KeyName"]
+        logger.error(f"{self._keyName}")
+        # self._keyName = list(instanceMetaData['public-keys'].keys())[0]
+        self._leaderPrivateIP = ec2_metadata.private_ipv4  # this is PRIVATE IP
+        self._tags = {k: v for k, v in (self.getLeader().tags or {}).items() if k != _TAG_KEY_TOIL_NODE_TYPE}
         # Grab the ARN name of the instance profile (a str) to apply to workers
-        self._leaderProfileArn = instanceMetaData['iam']['info']['InstanceProfileArn']
+        leader_info = ec2_metadata.iam_info
+        if leader_info is None:
+            logger.error("Failed to query leader IAM metadata. This may cause issues with the cluster in the future.")
```

Comment:
Can we be more specific about what kind of issues and why they might or might not happen? It's something like the leader might not have permissions to access AWS services, right?

Reply:
Also won't this always happen if the leader is not itself in AWS?

Reply:
`iam_info` may return None, which has information that is later used to provision new workers/nodes such as spot instances. `ec2_metadata.iam_info` returns None by catching a 404 response, so it is probably due to the leader not being on AWS or AWS is unreachable. I think it's probably better to wrap this in a retry with a timeout. If the metadata isn't reachable, there probably is a greater error, and Toil's provisioning will probably break.